src/content/articles/for-in-string-types.mdx

```diff
@@ -0,0 +1,212 @@
+The main difference is that `Object.keys()` only looks at properties on the object itself, while `for...in` loops also look at properties on the object's `prototype` chain.
+
+See [Workarounds](#workarounds) below for how you can get around this behavior in `Object.keys()` as well.
+
+## Workarounds
+
+Before you try to work around the design of the type checker, please pause a moment and reflect.
+The situations mentioned earlier in this post are real problems in JavaScript and TypeScript code that have caused countless bugs in software.
+
+**TypeScript is right to assume `string`s types.**
+
+### `Object` APIs
+
+TypeScript's type checking is designed to work with how JavaScript data typically flows and changes through an application.
+Needing to wrestle with the type checker is often a byproduct of application code going against safe JavaScript practices -- such as the aforementioned issues with `for...in` loops receiving unexpected key strings.
+In many cases, it might be good to listen to TypeScript and consider refactoring your code to use a JavaScript API better suited to the task.
+
+If all your code needs to do is get all the values in an object, an `Object.values()` API exists that returns those as an array.
+You can `for...of` loop over them instead:
+
+```ts twoslash
+// @lib: dom,esnext
+
+const counts = {
+  one: 1,
+  two: 2,
+};
+
+for (const [key, value] of Object.entries(counts)) {
+  //        ^?
+}
+```
+
+Alternately, if you do need keys, consider using the `Object.entries()` API instead.
+It returns a `[string, T][]` where `T` is the type of values in an object:
+
+```ts twoslash
+// @lib: dom,esnext
+
+const counts = {
+  one: 1,
+  two: 2,
+};
+
+for (const [key, value] of Object.entries(counts)) {
+  //        ^?
+}
+```
+
+Both `Object.entries()` and `Object.values()` allow you to write clean, idiomatic JavaScript without having to grapple with the type checker.
+
+### `Map`
+
+Taking an even larger step back, why is the application code looping over an object's keys and values?
+Is the object meant to be used as a general key-value store?
+
+If so: consider using the built-in `Map` data structure instead.
+It's made specifically for key-value stores.
+
+You can `for...of` over a `Map`'s values using `Array.from`:
+
+```ts twoslash
+// @lib: dom,esnext
+
+const counts = new Map([
+  ["one", 1],
+  ["two", 2],
+]);
+
+for (const [key, value] of Array.from(counts)) {
+  //        ^?
+}
+```
+
+`Map`s have the added benefit over `{}` objects of allowing any object to be a key without stringifying them first (which objects do).
+
+### Type Assertions
+
+Still, there are times when a `for...in` loop or `Object.keys()` might likely be mostly safe to use for object keys.
+If you are 100% absolutely completely sure that your use case won't crash from some bizarre type mismatch now or in the future, you can always use an `as` assertion to cast the loop variable to the right type.
```

Comment:
[Proofreading/nitpick] While mentioning "completely" and "absolutely" encourages devs to think deeply about the use case, I think saying "100% sure" conveys this point perfectly.  
```suggestion
If you are 100% sure that your use case won't crash from some bizarre type mismatch now or in the future, you can always use an `as` assertion to cast the loop variable to the right type.
```